### PR TITLE
ARROW-11741: [C++] Fix decimal casts on big endian platforms

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
cc @kiszk

**Comment by pitrou** [comment]
Green Travis-CI build: https://travis-ci.com/github/pitrou/arrow/builds/217951780

**Comment by pitrou** [comment]
Another possible fix would be to reorder Decimal fields based on endianness.

**Review by bkietz — CHANGES_REQUESTED** [review]
I think it makes most sense to reorder `BasicDecimal128::{low_bits_, high_bits_}` based on endianness. Furthermore, BasicDecimal256 should (eventually) follow suit by replacing `little_endian_array_` with `native_endian_array_`.

**Comment by pitrou** [comment]
Note that one issue with the `reinterpret_cast` approach is forcing unaligned loads. There may be some platforms around where that's unallowed (SPARC?).

**Comment by bkietz** [comment]
When would any of these elements not be aligned to a 16 byte boundary?

**Comment by pitrou** [comment]
That would be quite unlikely indeed :-)

**Comment by bkietz** [comment]
Isn't it impossible for conformant data?

**Comment by pitrou** [comment]
No, data can be backed by a buffer that wasn't allocated by Arrow and that's not aligned naturally. That said, it's certainly extremely unlikely to happen in practice, and I don't think we actually care in other places, so I'm fine with ignoring that possibility.

**Comment by pitrou** [comment]
As far as Decimal256 is concerned, though, adapting the code may be non-trivial to conform to a native-endian layout. And if we do it only for Decimal128, we get less orthogonality.

File: cpp/src/arrow/compute/kernels/codegen_internal.h [hunk 1]
```diff
@@ -663,13 +663,14 @@ struct ScalarUnaryNotNullStateful {
...
                      Datum* out) {
       ArrayData* out_arr = out->mutable_array();
-      auto out_data = out_arr->GetMutableValues<Decimal128>(1);
+      auto out_data = out_arr->GetMutableValues<uint8_t>(1, 0) + 16 * out_arr->offset;
       VisitArrayValuesInline<Arg0Type>(
           arg0,
           [&](Arg0Value v) {
-            *out_data++ = functor.op.template Call<OutValue, Arg0Value>(ctx, v);
+            functor.op.template Call<OutValue, Arg0Value>(ctx, v).ToBytes(out_data);
+            out_data += 16;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/codegen_internal.h:671]:
```suggestion
            functor.op.template Call<OutValue, Arg0Value>(ctx, v).ToBytes(out_bytes++->data());
```

File: cpp/src/arrow/compute/kernels/codegen_internal.h [hunk 2]
```diff
@@ -663,13 +663,14 @@ struct ScalarUnaryNotNullStateful {
...
           },
-          [&]() { ++out_data; });
+          [&]() { out_data += 16; });
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/codegen_internal.h:673]:
```suggestion
          [&]() { ++out_bytes; });
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Nothing else in the compute namespace assumes data may be unaligned, so supporting it only here is suspect to me. We should at least have warnings that buffers are unaligned; even on platforms where unaligned access is not an error it does decrease performance.

I've added https://issues.apache.org/jira/browse/ARROW-11748 to track refactoring Decimal* into native-endian ordering.

In the meantime, let's merge this with a slight modification to make the intent clearer:

**Comment by pitrou** [comment]
I applied your suggestions @bkietz , will merge. Thank you for the review!

**Comment by kiszk** [comment]
Thank you, I will watch ttps://issues.apache.org/jira/browse/ARROW-11748

**Pull request closed without merging**